drop-analytics/src/file_impl.rs

```diff
@@ -78,7 +79,13 @@ impl FileImpl {
         };
         events.push(event);
 
-        File::create(&self.event_path)?.write_all(serde_json::to_string(&events)?.as_bytes())?;
+        let payload = serde_json::to_string(&events)?;
+
+        // create unique temp path
+        let temp_path = format!("{}.tmp.{}", self.event_path, uuid::Uuid::new_v4());
+
+        File::create(&temp_path)?.write_all(payload.as_bytes())?;
+        std::fs::rename(temp_path, &self.event_path)?;
```

Comment:
Sure thing! I've started to notice weird logs from this place about invalid JSON. When printing it it was really invalid JSON, no idea how it could be reproduced until it hit me that when using tests that are doing a lot of writes to the database, it might be that several of these function calls get interlaced. Because the file is written in chunks, some chunks would be interlaced resulting in invalid JSON.

The usual way for such cases is to write to a unique temp file and then move it to the final destination(fs move is an atomic operation). However, even though this fixes this exact annoyance it will still render the file non-deterministic in that case as some stuff will be lost. The proper way would be to have a mutex around the writer as well but  it requires much more changes

Reply:
If that's the case then `Mutex` is a way to go IMHO. With you're solution it's possible to lose some events

Reply:
I think the current variant is still an improvement(no invalid json is produced), we can implement the mutexed solution once this causes us problems